hibernate-core/src/main/java/org/hibernate/dialect/function/SumReturnTypeResolver.java

```diff
@@ -66,6 +67,9 @@ public ReturnableType<?> resolveFunctionReturnType(
 			return impliedType;
 		}
 		final ReturnableType<?> argType = extractArgumentType( arguments, 1 );
+		if ( argType instanceof CustomType ) {
+			return argType;
+		}
```

Comment:
Instead of doing this, I was rather thinking about checking if the type has the desired target type already e.g.

```
			case SMALLINT:
			case TINYINT:
			case INTEGER:
			case BIGINT:
				return basicType.getJdbcJavaType() == Long.class ? basicType : longType;
			case FLOAT:
			case REAL:
			case DOUBLE:
				return basicType.getJdbcJavaType() == Double.class ? basicType : doubleType;
			case DECIMAL:
			case NUMERIC:
				return BigInteger.class.isAssignableFrom( basicType.getJdbcJavaType() )
								|| BigDecimal.class.isAssignableFrom( basicType.getJdbcJavaType() ) 
								? basicType 
								: bigDecimalType;
```